python/ray/ray_process_reaper.py

```diff
@@ -23,13 +24,23 @@ def sigterm_handler(*args):
         # Give a one-second grace period for other processes to clean up.
         time.sleep(SIGTERM_GRACE_PERIOD_SECONDS)
         # SIGKILL the pgroup (including ourselves) as a last-resort.
-        os.killpg(0, signal.SIGKILL)
+        if sys.platform == "win32":
+            atexit.unregister(sigterm_handler)
+            os.kill(0, signal.CTRL_BREAK_EVENT)
```

Comment:
Does this actually kill the whole "process group?" Also, I thought we don't need this reaper process at all in windows because we can just use the job object for fate sharing instead?

Reply:
[It does](https://docs.microsoft.com/en-us/windows/console/generateconsolectrlevent), but I think it only works for console applications. I'm working on better mechanisms for terminating programs later (which is mainly needed for more graceful shutdowns), but this was an easy port so I ported this too.

And yes, that's true, job objects obviate the need for the reaper. However:

- When debugging, I've found this more helpful than just killing all the children. Having the kernel immediately kill everything is the right choice for production, but during development it can interfere with debugging.

- This is still in the codebase and a theoretically possible code path (e.g. if the job objects can't be created for some reason), so it seemed better to make it do something useful. You might also want it later if people report issues and you want to have them try alternate mechanisms (e.g. to trigger shutdown behaviors more easily).

Reply:
If the reaper is not necessary for windows then I'd lean towards not including this change, but I don't feel that it's there's much harm here as this is still a very simple file. If it helps with debugging then I think it's fine to leave it in.

Reply:
Yeah it's definitely been helping me with debugging shutdown behavior, so it'd be nice if we can keep it for now. :) Thanks!